### PR TITLE
Avoid UnobservedTaskExceptions when getting an error in server to client streaming

### LINKED ISSUE #46169: Unobserved task exception in SignalR C# client when streaming
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When calling a streaming method on a server that throws an exception, the resulting HubException is handled on the client, but an additional unobserved exception remains on a task and is later processed by the UnobservedTaskException-Handler if one is registered.

I've looked briefly at the SignalR-code, can the issue be that in https://github.com/dotnet/aspnetcore/blob/51abb6ac8005a7715ee5c8b7134938c3cec87acb/src/SignalR/clients/csharp/Client.Core/src/HubConnectionExtensions.StreamAsChannelAsync.cs#L291-L307 the completion of the inputChannel is awaited in line 307, but WaitToReadAsync could already throw the exception in line 291, thus that await of the completion is not reached in such a case?

### Expected Behavior

No unobserved task exceptions should occur.

### Steps To Reproduce

Small repro here: https://github.com/MikeOnTea/SignalrUnobservedExceptionTest

Relevant client code is in Client/MainWindow.xaml.cs.

Steps to repro:
- Run the server
- Run the client
- Click the Test-Button in the Client that calls the streaming server method.
- A message box appears, showing the exception
- A second message box appears, showing an unobserved task exception got collected by the GC.

### Exceptions (if any)

_No response_

### .NET Version

7.0.101

### Anything else?

Tested with SignalR version 6.0.13 and briefly with 7.0.2.
Visual Studio version 2022 running on Windows 10.

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/46169

Filed https://github.com/dotnet/runtime/issues/80885 on the Runtime side since this is an unfortunate usage pattern.

### CODE REVIEW

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs [hunk 1]
```diff
@@ -603,13 +603,23 @@ public virtual IAsyncEnumerable<TResult> StreamAsyncCore<TResult>(string methodN
...
+            while (await reader.WaitToReadAsync(cts.Token).ConfigureAwait(false))
             {
-                yield return (T)item!;
+                while (reader.TryRead(out var item))
+                {
+                    yield return (T)item!;
+                }
             }
         }
+        finally
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:617]: Can't have a try catch around `yield` statements.

File: src/SignalR/clients/csharp/Client.Core/src/HubConnectionExtensions.StreamAsChannelAsync.cs [hunk 2]
```diff
@@ -314,6 +311,15 @@ private static async Task RunChannel<TResult>(ChannelReader<object?> inputChanne
         {
             // This will safely no-op if the catch block above ran.
             outputChannel.Writer.TryComplete();
+
+            try
+            {
+                // Needed to avoid UnobservedTaskExceptions
+                await inputChannel.Completion.ConfigureAwait(false);
```
Comment (davidfowl) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnectionExtensions.StreamAsChannelAsync.cs:318]: https://github.com/dotnet/runtime/issues/80885#issuecomment-1398220271 feels cleaner.

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs [hunk 3]
```diff
@@ -603,13 +603,23 @@ public virtual IAsyncEnumerable<TResult> StreamAsyncCore<TResult>(string methodN
...
+        {
+            // If WaitToReadAsync threw, this will replace that and throw the same Exception
+            // Needed to avoid UnobservedTaskExceptions
+            await reader.Completion.ConfigureAwait(false);
```
Comment (theodorzoulias) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:621]: Is it guaranteed that the `reader.Completion` task is completed at this point?

**Review by davidfowl — APPROVED** [review]

**Pull request merged**